pkg/R/RDD.R

```diff
@@ -98,48 +138,44 @@ setMethod("getJRDD", signature(rdd = "PipelinedRDD"),
+                                         ascii = TRUE)
 
-            refs <- lapply(ls(.broadcastNames),
-                           function(name) { get(name, .broadcastNames) })
-            broadcastArr <- .jarray(refs,
-                                    "org/apache/spark/broadcast/Broadcast")
+            broadcastArr <- lapply(ls(.broadcastNames),
+                                   function(name) { get(name, .broadcastNames) })
 
             depsBin <- getDependencies(computeFunc)
-            depsBinArr <- .jarray(depsBin)
 
             prev_jrdd <- rdd@prev_jrdd
 
             if (dataSerialization) {
-              rddRef <- new(J("edu.berkeley.cs.amplab.sparkr.RRDD"),
-                             prev_jrdd$rdd(),
-                             serializedFuncArr,
-                             rdd@env$serialized,
-                             depsBinArr,
-                             packageNamesArr,
-                             as.character(.sparkREnv[["libname"]]),
-                             broadcastArr,
-                             prev_jrdd$classTag())
+              rddRef <- newJObject("edu.berkeley.cs.amplab.sparkr.RRDD",
+                                   callJMethod(prev_jrdd, "rdd"),
+                                   serializedFuncArr,
+                                   rdd@env$serialized,
+                                   depsBin,
+                                   packageNamesArr,
+                                   as.character(.sparkREnv[["libname"]]),
+                                   broadcastArr,
+                                   callJMethod(prev_jrdd, "classTag"))
             } else {
-              rddRef <- new(J("edu.berkeley.cs.amplab.sparkr.StringRRDD"),
-                             prev_jrdd$rdd(),
-                             serializedFuncArr,
-                             rdd@env$serialized,
-                             depsBinArr,
-                             packageNamesArr,
-                             as.character(.sparkREnv[["libname"]]),
-                             broadcastArr,
-                             prev_jrdd$classTag())
+              rddRef <- newJObject("edu.berkeley.cs.amplab.sparkr.StringRRDD",
+                                   callJMethod(prev_jrdd, "rdd"),
+                                   serializedFuncArr,
+                                   rdd@env$serialized,
+                                   depsBin,
+                                   packageNamesArr,
+                                   as.character(.sparkREnv[["libname"]]),
+                                   broadcastArr,
+                                   callJMethod(prev_jrdd, "classTag"))
             }
             # Save the serialization flag after we create a RRDD
             rdd@env$serialized <- dataSerialization
-            rdd@env$jrdd_val <- rddRef$asJavaRDD()
+            rdd@env$jrdd_val <- callJMethod(rddRef, "asJavaRDD") # rddRef$asJavaRDD()
```

Comment:
So the convention we have in the code is that we save a pointer to the JavaRDD in the R side. This is because we call methods like `collectPartitions` etc which are (a) sometimes only present in the JavaRDD and (b) the type signatures we depend on are the java versions. So we need the `asJavaRDD` to remain consistent with this